packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,236 @@
+    //                      by not setting the `totalRewardsDistributed` correctly to 101).
+    //     - Window 2 Tree: Owner sets `totalRewardsDistributed=1` and insert proofs that give
+    //                      claimant A 1 token. The owner correctly set `totalRewardsDistributed` this time.
+    //     - At this point contract owns 100 + 1 = 101 tokens. Now, imagine the following sequence:
+    //       (1) Claimant A claims 50 tokens for Window 1, contract now has 101 - 50 = 51 tokens.
+    //       (2) Claimant B claims 51 tokens for Window 1, contract now has 51 - 51 = 0 tokens.
+    //       (3) Claimant A tries to claim 1 token for Window 2 but fails because contract has 0 tokens.
+    //     - In summary, the contract owner created a race for step(2) and step(3) in which the first
+    //       claim would succeed and the second claim would fail, even though both claimants would expect
+    //       their claims to suceed.
+    function setWindowMerkleRoot(
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        uint256 indexToSeed = lastSeededIndex;
+        lastSeededIndex = indexToSeed.add(1);
+
+        _seedWindow(indexToSeed, totalRewardsDistributed, windowStart, rewardToken, merkleRoot);
+    }
+
+    // When `window.locked` is true, claims are blocked for that window.
+    function setWindowLock(uint256 windowIndex, bool lockValue) external nonReentrant() onlyOwner {
+        merkleWindows[windowIndex].locked = lockValue;
+        emit SetWindowLock(msg.sender, windowIndex, lockValue);
+    }
+
+    // Overwrite merkle root for specified window index. Owner must deposit `totalRewardsDistributed`
+    // into contract.
+    // TODO: Should we require that `windowIndex <= lastSeededIndex` or should this function just be
+    // a generalized `setWindowMerkleRoot` whereby the owner can set a merkle root for any index?
+    function resetWindowMerkleRoot(
+        uint256 windowIndex,
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        _seedWindow(windowIndex, totalRewardsDistributed, windowStart, rewardToken, merkleRoot);
+    }
+
+    // Emergency methods to transfer rewards into and out of the contract
+    // incase the contract was configured improperly.
+    function withdrawRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransfer(msg.sender, amount);
+        emit WithdrawRewards(msg.sender, amount);
+    }
+
+    function depositRewards(address rewardCurrency, uint256 amount) external nonReentrant() onlyOwner {
+        IERC20(rewardCurrency).safeTransferFrom(msg.sender, address(this), amount);
+        emit DepositRewards(msg.sender, amount);
+    }
```

Comment:
nit: just want to double check that we _want_ an explicit method for this as opposed to just doing raw transfers to the contract?

Reply:
I am not strongly opinionated actually. It's simpler to just go with raw transfers

Reply:
Removed this function